components/sup/src/manager/user_config_watcher.rs

```diff
@@ -0,0 +1,251 @@
+pub trait Serviceable {
+    fn name(&self) -> &String;
+    fn path(&self) -> &PathBuf;
+    fn service_group(&self) -> &ServiceGroup;
+}
+
+impl Serviceable for Service {
+    fn name(&self) -> &String {
+        &self.pkg.name
+    }
+
+    fn path(&self) -> &PathBuf {
+        &self.pkg.svc_path
+    }
+
+    fn service_group(&self) -> &ServiceGroup {
+        &self.service_group
+    }
+}
+
+
+pub struct UserConfigWatcher {
+    // TODO use a HashSet? we don't really need the boolean:
+    // if the key is present, it's true, otherwise it's false.
+    states: HashMap<String, Arc<AtomicBool>>,
+}
+
+impl UserConfigWatcher {
+    pub fn new() -> Self {
+        Self { states: HashMap::new() }
+    }
+
+    pub fn add<T: Serviceable>(&mut self, service: &T) {
+        self.states.entry(service.name().clone()).or_insert_with(
+            || {
+                let have_events = Arc::new(AtomicBool::new(false));
+                Worker::run(&service.path(), Arc::clone(&have_events));
+
+                have_events
+            },
+        );
+        outputln!(preamble service.service_group(), "Watching {}", USER_CONFIG_FILE);
+    }
+
+    pub fn have_events_for<T: Serviceable>(&self, service: &T) -> bool {
+        if let Some(val) = self.states.get(service.name()) {
+            let val = val.load(Ordering::Relaxed);
+
+            val
+        } else {
+            false
+        }
+    }
+
+    pub fn reset_events_for<T: Serviceable>(&mut self, service: &T) {
+        if let Some(val) = self.states.get_mut(service.name()) {
+            val.store(false, Ordering::Relaxed);
+        }
+    }
+}
+
+struct UserConfigCallbacks {
+    have_events: Arc<AtomicBool>,
+}
+
+impl Callbacks for UserConfigCallbacks {
+    fn file_appeared(&mut self, _: &Path) {
+        self.have_events.store(true, Ordering::Relaxed);
+    }
+
+    fn file_modified(&mut self, _: &Path) {
+        self.have_events.store(true, Ordering::Relaxed);
+    }
+
+    fn file_disappeared(&mut self, _: &Path) {
+        self.have_events.store(true, Ordering::Relaxed);
+    }
+}
+
+struct Worker;
+
+impl Worker {
+    // starts a new thread with the file watcher on the service's user.toml file
+    pub fn run(service_path: &Path, have_events: Arc<AtomicBool>) {
+        let path = service_path.join(USER_CONFIG_FILE);
+        Self::setup_watcher(path, have_events);
+    }
+
+    fn setup_watcher(path: PathBuf, have_events: Arc<AtomicBool>) {
+        ThreadBuilder::new()
+            .name(format!("user-config-watcher-{}", path.display()))
+            .spawn(move || {
+                debug!(
+                    "UserConfigWatcher({}) worker thread starting",
+                    path.display()
+                );
+                let callbacks = UserConfigCallbacks { have_events: have_events };
+                let mut file_watcher =
+                    default_file_watcher(&path, callbacks).expect("creating user-config watcher");
```

Comment:
Instead of using `expect`, I'd use `outputln!` macro to report an error and maybe quit the thread. I'm not sure if we should panic in this case.

Reply:
Good point, I've committed a change.